randovania/game_connection/connector/cs_remote_connector.py

```diff
@@ -0,0 +1,215 @@
+
+    async def _update_collected_indices(self) -> None:
+        self.game.region_list.ensure_has_node_cache()
+        indices = sorted(self.game.region_list._pickup_index_to_node)
+        flag_nums = [INDICES_FLAGS_START + pickup_index.index for pickup_index in indices]
+        flags = await self.executor.get_flags(flag_nums)
+
+        for index, collected in zip(indices, flags, strict=True):
+            if collected:
+                self.PickupIndexCollected.emit(index)
+
+    async def _update_inventory(self) -> None:
+        new_inventory = await self.get_inventory()
+        if new_inventory != self.last_inventory:
+            self.last_inventory = new_inventory
+            self.InventoryUpdated.emit(new_inventory)
+
+    async def get_inventory(self) -> Inventory:
+        inventory = {}
+
+        item_db = self.game.resource_database.item
+        get_item = self.game.resource_database.get_item
+
+        # Normal items
+        normal_items = [item for item in item_db if "flag" in item.extra]
+        flag_nums = [item.extra["flag"] for item in normal_items]
+        flags = await self.executor.get_flags(flag_nums)
+
+        for item, flag in zip(normal_items, flags, strict=True):
+            if flag:
+                inv = InventoryItem(1, 1)
+            else:
+                inv = InventoryItem(0, 0)
+            inventory[item] = inv
+
+        # Puppies
+        puppy_flags = await self.executor.get_flags(PUPPIES_FLAGS)
+        puppies = get_item("puppies")
+        puppy_count = sum(1 for pup in puppy_flags if pup)
+        inventory[puppies] = InventoryItem(puppy_count, puppy_count)
+
+        # Missile ammo
+        weapons = await self.executor.get_weapons()
+        missile_ammo = get_item("missile")
+        missiles = next((weapon for weapon in weapons if weapon.weapon_id in MISSILE_IDS), None)
+        if missiles is not None:
+            inventory[missile_ammo] = InventoryItem(missiles.ammo, missiles.capacity)
+        else:
+            inventory[missile_ammo] = InventoryItem(0, 0)
+
+        # Life
+        life = await self.executor.read_memory(0, 2, base_offset="current_hp")
+        life += await self.executor.read_memory(0, 2, base_offset="max_hp")
+        hp, max_hp = struct.unpack("<2h", life)
+        inventory[get_item("lifeCapsule")] = InventoryItem(hp, max_hp)
+
+        return Inventory(inventory)
+
+    async def _receive_items(self) -> None:
+        if not self.remote_pickups:
+            return
+
+        num_pickups = await self.executor.get_received_items()
+
+        # no new pickups to send
+        if num_pickups >= len(self.remote_pickups):
+            return
+
+        # pickup has been received by the game:
+        # increment received pickups and restart the cycle
+        if await self.executor.get_flag(ITEM_RECEIVED_FLAG):
+            await self.executor.set_received_items(num_pickups + 1)
+            await self.executor.set_flag(ITEM_RECEIVED_FLAG, False)
```

Comment:
How well does this handle if connection is lost between the previous line and this one?

Reply:
It can cause problems, but it would cause problems in any order and it's such an extreme corner case (these awaits take virtually no time at all to execute) that I don't think it's worth worrying about